### PR TITLE
feat: add interceptor for error responses

### PR DESCRIPTION
**Pull request by mojito317**
## PR summary

Add interceptor for error response augmentation

Fixes: s1003

## PR Checklist

Please make sure that your PR fulfills the following requirements:

- [x] The commit message follows the
[Angular Commit Message Guidelines](https://github.com/angular/angular/blob/master/CONTRIBUTING.md#-commit-message-guidelines).
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type  

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] New tests
- [ ] Build/CI related changes
- [ ] Documentation content changes
- [ ] Other (please describe)

## What is the current behavior?
Cloudant/CouchDB error models are handled by the core with only the `error` field being processed. Obtaining e.g. the `reason` field requires additional user code.

## What is the new behavior?
The error response is augmented with `errors` and `trace` properties.
This improves the way the core handles the error response and enhances the user facing exception string to include the `reason`.
The `trace` is the CouchDB request ID which can be used to correlate with logs.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No
The existing error model fields are retained, the new fields are backwards compatible.

## Other information

Add `errorResponseInterceptor`.
Modify `cloudantBaseService` to apply interceptor.
New tests in `errorResponseInterceptorTest`.
Fix service options on `test/unit/features/testDataProviders.js`.
Add `nock` for testing.

### CODE REVIEW

File: test/unit/errorResponseInterceptor.test.js [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+    assert.deepStrictEqual(responseResult, test.expectedResponse);
+    if (test.expectedResponseMessage) {
+      // check message includes SyntaxError for edge case test_no_augment_invalid_json:
+      assert.ok(responseMessage.includes(test.expectedResponseMessage));
+      assert.ok(!responseStack.includes('errorResponseInterceptor'));
+    }
+  });
+
+  it('test_added', async () => {
+    const expectedNumberOfResponseInterceptors = numberOfBaseInterceptors + 1;
```
Comment (ricellis) [inline test/unit/errorResponseInterceptor.test.js:366]: Since `numberOfBaseInterceptors = 1` the expected here is `2`.
AFAIK if retries are not also configured the error interceptor should be the only one, so I don't really understand why we're asserting `2`. (Except that possibly we're accidentally adding 2 when we shouldn't be because of the `newInstance`/`configureService` thing.

**Review by ricellis — COMMENTED** [review]
I'm not sure about the new ensure hook added if client is re-set commit - seems like Node core might be different in this respect.

File: lib/errorResponseInterceptor.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+export function errorResponseInterceptor(axiosError) {
+  if (
```
Comment (eiri) [inline lib/errorResponseInterceptor.ts:18]: I am confused on inconsistent use of optional chain on `axiosError` in this `if`. Sometimes it used, sometime it is missing and anyway, if we expect `axiosError` to be able to be `nil`, it'd be cleaner to check and return promise reject as a first check and then change condition to use it with confidence that it is present.

File: lib/errorResponseInterceptor.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    axiosError?.response && // must have a response
+    axiosError?.response.status >= 400 &&
+    axiosError?.config.method !== 'head' &&
```
Comment (eiri) [inline lib/errorResponseInterceptor.ts:21]: This seems redundant to below check on data presence. Do we expect a case where we have "head" req with body?

File: lib/errorResponseInterceptor.ts [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    // must have data:
+    axiosError.response.data &&
+    // must be a JSON:
+    axiosError?.config.responseType === 'json' &&
```
Comment (eiri) [inline lib/errorResponseInterceptor.ts:25]: This again seems redundant. What does this give us that not covered by further checks on content type and instance of Object? Or other way around - if this passes, can we skip next two checks?

File: lib/errorResponseInterceptor.ts [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+      };
+      if (axiosError.response.data.reason) {
+        error.message += `: ${axiosError.response.data.reason}`;
+      }
+      // Add the new error as part of an errors array.
+      axiosError.response.data.errors = [error];
+    }
+    if (axiosError.response.data.errors) {
+      // Map x-couch-request-id if available to the trace field
+      const trace = axiosError?.response?.headers['x-couch-request-id'];
```
Comment (eiri) [inline lib/errorResponseInterceptor.ts:48]: Given that we accessed `axiosError.response` plenty before those optional chaining looks unnecessary

File: lib/cloudantBaseService.ts [hunk 6]
```diff
@@ -142,6 +149,11 @@ export default abstract class CloudantBaseService extends BaseService {
     // Read external configuration and set as request defaults.
     super.configureService(serviceName);
     this.configureSessionAuthenticator();
+    // Add response interceptor for error transforms
+    this.getHttpClient().interceptors.response.use(
+      (response) => response,
+      (axiosError) => errorResponseInterceptor(axiosError)
+    );
```
[the remark this answers is not in the document]
Comment (mojito317) [reply lib/cloudantBaseService.ts:156]: `configureService` overwrites the initial interceptor settings with `super.configureService(serviceName)` loosing the interceptor we have initially added in the constructor. Similarly to the session authenticator config this needs to be called here as well. 

The test case of `test_added_via_configureService` simulates what's happening when `new CloudantV1` then `configureService` happens (as in `newInstance`) so we can keep that test case, I think, but I can add another test where we call `newInstance` instead of `new Cloudant` if you insist.

File: lib/errorResponseInterceptor.ts (hunk 5 above)
Comment (mojito317) [reply lib/errorResponseInterceptor.ts]: 3453e093.

File: lib/errorResponseInterceptor.ts (hunk 2 above)
Comment (mojito317) [reply lib/errorResponseInterceptor.ts:18]: I left optionals by a coincidence. Personally, I do not like using them, those were used in the proposal, so I think we should ask @ricellis' opinion about this. I do not think an `axiosError` would be `nil`, so I would not cover that, but LMK if I miss an edge case here.

Anyways, I got rid off the optionals in 3453e093.

File: test/unit/errorResponseInterceptor.test.js (hunk 1 above)
Comment (mojito317) [reply test/unit/errorResponseInterceptor.test.js:366]: I added one comment for this in 95ebee7b.

File: lib/cloudantBaseService.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (ricellis) [reply lib/cloudantBaseService.ts:156]: Oh I see it now via the fact that the `configureService` creates a new request wrapper with a new client.
Probably the test is fine, having an extra `newInstance` as well might protect us against divergence of that implenentation from what is simulated in the existing test, but I think we test `newInstance` in other paths so we're probably OK.

File: test/unit/errorResponseInterceptor.test.js (hunk 1 above)
Comment (ricellis) [reply test/unit/errorResponseInterceptor.test.js:366]: I think I probably even did the PR for that at some point and forgetten about it.

File: lib/errorResponseInterceptor.ts (hunk 2 above)
Comment (ricellis) [reply lib/errorResponseInterceptor.ts:18]: Yeah it's fine, the prototype was a quick experiment.

**Review by ricellis — APPROVED** [review]
